File: src/frontend/src/lib/features/recipes/suggestionReason.ts
import { m } from '$shell/i18n';

import type { Suggestion } from './types';

/**
 * The one quiet line under a suggestion's title, in the reader's language.
 *
 * The server sends a code and at most a subject, never prose: the wording
 * belongs here because this is what knows which of two languages somebody
 * reads, and because a sentence on the wire cannot be translated after it
 * arrives.
 *
 * Null whenever no single signal decided the ranking. That is an ordinary
 * answer and it means show nothing — a good suggestion with no explanation is
 * fine, and an invented one is a lie that discredits the reasons that were
 * true.
 */
export function reasonLineFor(suggestion: Suggestion): string | null {
  const reason = suggestion.reason;

  if (!reason) {
    return null;
  }

  switch (reason.code) {
    case 'affinity':
      return m['suggestions.reason.affinity']();
    case 'rediscovery':
      return m['suggestions.reason.rediscovery']();
    case 'season':
      return m['suggestions.reason.season']();
    case 'slot':
      return m['suggestions.reason.slot']();
    case 'fresh':
      return m['suggestions.reason.fresh']();
    case 'similar':
      return m['suggestions.reason.similar']();
    // The three that name something. The server already refuses to send one of
    // these without a subject, and the guard is here anyway: "You often cook"
    // with a hole in it is worse than no line at all, and a contract is a thing
    // that can change.
    case 'tag':
      return reason.subject ? m['suggestions.reason.tag']({ subject: reason.subject }) : null;
    case 'ingredient':
      return reason.subject
        ? m['suggestions.reason.ingredient']({ subject: reason.subject })
        : null;
    case 'household':
      return reason.subject ? m['suggestions.reason.household']({ subject: reason.subject }) : null;
    default:
      // A code this client does not know yet. Newer server, older app: say
      // nothing rather than the code.
      return null;
  }
}

/**
 * The reason lines for a shortlist walked one panel at a time, each null where
 * it would only repeat the panel before it.
 *
 * Every repeat is still true — a household that cooks its favourites really
 * does get "one you keep coming back to" five times. But read in a row, five
 * copies of the same fact stop reading as a fact and start reading as
 * boilerplate, which undoes what a reason is for. So the set is explained
 * rather than each member of it: the first of a run says why, and the rest fall
 * back to the panel's fixed line.
 *
 * Compared as rendered lines rather than codes, so "you often cook soup" and
 * "you often cook curry" are two different things to say, which they are.
 */
export function reasonLinesFor(suggestions: readonly Suggestion[]): (string | null)[] {
  const lines = suggestions.map(reasonLineFor);

  return lines.map((line, index) => (index > 0 && line === lines[index - 1] ? null : line));
}

File: src/backend/src/Domain/Suggestions/SuggestionReason.cs
namespace Domain.Suggestions;

/// <summary>
/// Why one recipe was suggested, named after the term that actually decided it.
/// </summary>
/// <remarks>
/// <para>
/// A reason is never composed after the fact. The ranker returns a score broken
/// into its terms and the reason is whichever term dominated — so "you haven't
/// made this since April" means the rediscovery term was large, not that a
/// sentence was written to fit a recipe picked for other reasons.
/// </para>
/// <para>
/// When no single term dominates there is <see cref="None"/> and the card shows
/// its ordinary meta line. An unexplained good suggestion is fine; an invented
/// explanation is a lie, and catching one discredits every reason that was true.
/// </para>
/// </remarks>
public enum SuggestionReason
{
    /// <summary>No term dominated. Say nothing.</summary>
    None = 0,

    /// <summary>They cook this one. The card already carries the count.</summary>
    Affinity = 1,

    /// <summary>Liked once, and not made for a long time.</summary>
    Rediscovery = 2,

    /// <summary>It carries a tag they cook a lot of. The tag is the subject.</summary>
    Tag = 3,

    /// <summary>It uses an ingredient they cook a lot with. The name is the subject.</summary>
    Ingredient = 4,

    /// <summary>This household cooks this kind of thing at this time of year.</summary>
    Season = 5,

    /// <summary>It is usually planned in the slot that was asked for.</summary>
    Slot = 6,

    /// <summary>Somebody else in the household cooks it. Their name is the subject.</summary>
    Household = 7,

    /// <summary>New in the book, and nobody has cooked it yet.</summary>
    Fresh = 8,

    /// <summary>It resembles the recipe being looked at.</summary>
    Similar = 9
}
